#137 Add Google Analytics to Briefcase

jpknox opened the pull request:
Add a checkBox in the settings tab so the user can give or deny their consent to having their behaviour tracked. The setting is persisted.

Generate a UUID for each unique user. The UUID is persisted.

On startup track the following: App name (Briefcase), app version, operating system type, unique user ID.

jpknox commented on build.gradle:
```diff
@@ -75,6 +76,7 @@ buildConfig {
     version = getVersionName()
     clsName = 'BuildConfig'
     packageName = 'org.opendatakit.briefcase.buildconfig'
+    buildConfigField 'String', 'GOOGLE_TRACKING_ID', "UA-102618497-1"
```
This ID is just one I've made for testing purposes.

jpknox commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  /**
+   * A helper class used to streamline the creation Events.
```
Typo: "creation Events" should be "creation of Events".

dcbriccetti replied:
Since build() returns an ```EventHit```, maybe it should read “... creation of events” or “... creation of EventHits”.

jpknox replied:
Nicely spotted, I'll change it to "creation of EventHits".

jpknox commented on src/org/opendatakit/briefcase/model/BriefcasePreferences.java:
```diff
@@ -181,4 +185,40 @@ public static HttpHost getBriefCaseProxyConnection() {
+  }
+
+  /**
+   * Get the user's persisted decision to being tracked.
```
Grammar: should be more a long the lines of "Get the user's persisted decision regarding their consent to being tracked".

jpknox commented on src/org/opendatakit/briefcase/model/BriefcasePreferences.java:
```diff
@@ -181,4 +185,40 @@ public static HttpHost getBriefCaseProxyConnection() {
+   */
+  public static String getUniqueUserID() {
+    String userIdErrorMessage = "An error occurred generating the UUID.";
+    String uniqueUserID = Preference.APPLICATION_SCOPED.get(BriefcasePreferences.BRIEFCASE_UNIQUE_USER_ID_PROPERTY, userIdErrorMessage);
```
This line along with 217 should be refactored to make it clearer that the Preferences will return the userIdErrorMessage value only if no UUID already exists. At the moment it looks like some kind of obscure critical error is being caught.

jpknox commented on src/org/opendatakit/briefcase/ui/SettingsPanel.java:
```diff
@@ -1,4 +1,4 @@
-package org.opendatakit.briefcase.ui;
+    package org.opendatakit.briefcase.ui;
```
I'll undo this.

jpknox commented on src/org/opendatakit/briefcase/ui/SettingsPanel.java:
```diff
@@ -229,6 +241,22 @@ public void actionPerformed(ActionEvent e) {
+    public class TrackingConsentToggleListener implements ActionListener {
+        @Override
+        public void actionPerformed(ActionEvent e) {
+            System.out.println("Tracking consent checkbox toggled.");
```
I'll remove this line.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+   *
+   * <P>An {@link EventHit} object is used to model the information being tracked.</P>
+   */
+  private static class EventBuilder {
```
It builds EventHit objects, so maybe call it EventHitBuilder?

jpknox replied:
Nice spot once again, I'll amend it.

dcbriccetti commented on src/org/opendatakit/briefcase/ui/MessageStrings.java:
```diff
@@ -81,5 +81,6 @@
       "If you are behind a proxy, try setting up your proxy details through 'Settings' tab.";
 
   public static final String PARALLEL_PULLS = "Pull submissions in parallel (experimental)";
+  public static final String TRACKING_CONSENT = "Allow tracking of non-personal usage data.";
```
The other lines don’t end with a period.

jpknox replied:
I swore I'd double checked this but you're right, my mistake :p

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+  private static final String UNIQUE_USER_ID = BriefcasePreferences.getUniqueUserID();
+  private static final String EVENT_CATEGORY = "ODK Briefcase";
+
+  private static BriefcaseAnalytics instance = null;
```
The redundant initializers here are noise that slow down reading. I suggest:
```
  private static BriefcaseAnalytics instance;
  private static GoogleAnalytics googleAnalytics;
  private static final GoogleAnalyticsConfig config = new GoogleAnalyticsConfig();
  private static boolean hasTrackedStartup;
```
And ```config``` can be initialized here and made final.

jpknox replied:
Good point. I was hesitant to set `config` as final because I needed to make changes within it, but I later learned how final only stops the variable from being redeclared, it doesn't stop the developer from making changes within the object itself.

dcbriccetti replied:
Also noticed I recommend you change that Boolean to boolean. No need to use a reference type when a primitive will do. This boolean only needs two states, not three.

jpknox replied:
You have a good point so I tried to switch them to boolean, where I found that BriefcasePreferences insists on storing things as Strings, which is much easier if I use Boolean instead, otherwise I'd be switching from Boolean to boolean which will make things complicated. It's getting a bit late here so I'm going to leave them as Boolean.

dcbriccetti replied:
I was referring to ```hasTrackedStartup```, which is not used in the way you describe.

dcbriccetti replied:
setBriefcaseTrackingConsentProperty(Boolean value) takes a Boolean, and if you pass it a boolean, “auto boxing” will do the right thing.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+   * <P>The constructor is private so as to adhere to the rules of the Singleton pattern.</P>
+   */
+  private BriefcaseAnalytics() {
+    config = new GoogleAnalyticsConfig();
```
This line no longer needed.

jpknox replied:
Will remove.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+  private BriefcaseAnalytics() {
+    config = new GoogleAnalyticsConfig();
+    checkTrackingConsentState();
+    this.googleAnalytics = new GoogleAnalytics(config, GOOGLE_TRACKING_ID);
```
Remove ```this.```. Good style says to not access static elements via an instance. (Are you using IDEA? It highlights these things.)

jpknox replied:
I must have missed them. You're right about how I shouldn't refer to class-wide variables using the context of the current instance - will fix.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+   * @return the instantiated Singleton of {@link BriefcaseAnalytics}.
+   */
+  public static BriefcaseAnalytics getInstance() {
+    return instance == null ? new BriefcaseAnalytics() : instance;
```
```instance``` always remains null. I suggest
```
    if (instance == null) {
      instance = new BriefcaseAnalytics();
    }
    return instance;
```

dcbriccetti replied:
Here’s how I discovered that ```instance``` was never set (other than being redundantly set to null via its initializer): I removed that initializer, then IDEA told me immediately that it was never set. So, thanks, IDEA!

jpknox replied:
That's pretty cool, Eclipse was good albeit my time with it has helped me to really appreciate IntelliJ.

jpknox replied:
Thanks for finding the bug; I've gone for `return instance == null ? (instance = new BriefcaseAnalytics()) : instance;`

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+   * <P>Note that the method will check to see if it has already tracked this session's startup.</P>
+   */
+  public void trackStartup() {
+    if (hasTrackedStartup == false) {
```
Reduce noise with ```!hasTrackedStartup```.

jpknox replied:
Your suggestion is clearer, will do.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+  private static BriefcaseAnalytics instance = null;
+  private static GoogleAnalytics googleAnalytics = null;
+  private static GoogleAnalyticsConfig config = null;
+  private static Boolean briefcaseTrackingConsent = null;
```
briefcaseTrackingConsent  doesn’t need to be a field.

jpknox replied:
I like how it improves readability especially within checkStateTrackingConsent().

dcbriccetti replied:
OK, how about you use it there, but as a variable local to the method (not a field)?

jpknox replied:
My goal is to persist its state across the whole life cycle of the application, so a accidental second call to it in the app won't double the number of startups logged. I believe that if it is declared in the method it won't persist to the next time the method is called - I may be mistaken on this.

dcbriccetti replied:
Again, IDEA—alerted me that it doesn’t need to be a field. Meaning you weren’t doing anything with it that required its value being remembered outside that use in a single function.

jpknox replied:
I'm still not sure it will stop a second startup being tracked if I remove the field and call it twice. The Boolean variable 'hasTrackedStartup' must have its state persisted across each method call in order to stop a developer accidentally tracking unnecessary startup events.

dcbriccetti replied:
In my modified version of your code (which you might look at), I have this:
```
  private static boolean hasTrackedStartup;
```
It does need to be a field. ```briefcaseTrackingConsent``` doesn’t.

jpknox replied:
My mistake, I just realised you were talking about `briefcaseTrackingConsent`.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+   * in case the user has unticked the consent checkbox within the session.</P>
+   */
+  private void checkTrackingConsentState() {
+    briefcaseTrackingConsent = BriefcasePreferences.getBriefcaseTrackingConsentProperty();
```
With the field removed, these two lines can become
```    config.setGatherStats(BriefcasePreferences.getBriefcaseTrackingConsentProperty());```

jpknox replied:
I think your comment was meant for line 253 in 'SettingsPanel.java', good point though - will do.

dcbriccetti replied:
No, I meant here.
```
  private void checkTrackingConsentState() {
    config.setGatherStats(BriefcasePreferences.getBriefcaseTrackingConsentProperty());
  }
```
You can always look at https://github.com/dcbriccetti/briefcase/commit/ac98cfe2fc31b080aabe30225b1688775af1866e to see all my recommended changes, made.

jpknox replied:
Oh, I see what you mean. I like how my way makes the method clearer to read but I prefer how your way makes the class look clearer overall by decreasing the number of fields - I'll switch to your way.

I'm grateful that you made the changes, but I find that I learn a lot better by doing the changes myself. I'm going to keep doing the changes manually, for now at least.

dcbriccetti replied:
By all means, yes. I offer it to you if you want to see that the result does, in fact, work, and to make clearer anything I missed with the individual comments. 

I love Git and how I can get a local copy of a pull request and run and experiment with it myself.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+   */
+  private static class EventBuilder {
+
+    private String trackingID = null;
```
All of these redundant ``` = null``` initializers slow reading and add no value.

jpknox replied:
I'll remove them. Once in the past I tried to initialise a variable within a try-catch black and it insisted that I declare the variable as null before the 'try' section, ever since then I've gotten in to the habit of declaring absolutely everything null - I may rethink this :).

dcbriccetti replied:
This may bolster your confidence:
https://docs.oracle.com/javase/specs/jls/se7/html/jls-4.html#jls-4.12.5

jpknox replied:
I'll remember that, thanks for the tip 👍

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcasePreferences.java:
```diff
@@ -181,4 +185,40 @@ public static HttpHost getBriefCaseProxyConnection() {
+   * @param value (required) the boolean value representing the user's decision.
+   */
+  public static void setBriefcaseTrackingConsentProperty(Boolean value) {
+    if (value == null) {
```
Collapse four lines into:
```BriefcasePreferences.setBriefcaseTrackingConsentProperty(chkTrackingConsent.isSelected());```

jpknox replied:
Good point, will do.

jpknox commented on src/org/opendatakit/briefcase/ui/SettingsPanel.java:
```diff
@@ -83,14 +85,20 @@ public SettingsPanel(MainBriefcaseWindow parentWindow) {
         chkParallel.setSelected(BriefcasePreferences.getBriefcaseParallelPullsProperty());
         chkParallel.addActionListener(new ParallelPullToggleListener());
 
+        lblTrackingConsent = new JLabel(MessageStrings.TRACKING_CONSENT);
```
I'd hoped it would as straightforward as giving each label an ActionListener, but they don't have a method to set one as such.

dcbriccetti replied:
I would have used this constructor
https://docs.oracle.com/javase/7/docs/api/javax/swing/JCheckBox.html#JCheckBox(java.lang.String)
to provide the label. Then clicking on the label would toggle the checkbox. You did the right thing, probably, by continuing in the existing pattern.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,182 @@
+
+    public EventHit build() {
+      EventHit eventHit = new EventHit();
+      eventHit.trackingId(this.trackingID);
```
All the ```this.```s here add no value.

jpknox replied:
I'll remove them.

jpknox commented:
I've updated my PR in response to @dcbriccetti's feedback.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,179 @@
+  private static final String APPLICATION_VERSION = BuildConfig.VERSION;
+
+  private static BriefcaseAnalytics instance;
+  private static final GoogleAnalyticsConfig CONFIG = new GoogleAnalyticsConfig();
```
The all-caps suggests that config is a constant, but it isn’t.

jpknox replied:
I got confused by the difference between constants and final references.
`static final int YEAR = 2017` would be a constant because of how (I think) it's a primitive type.
'static final MyClass myClass' would just be a final reference because of how it's a complex type which cannot be reassigned to some other instance/object.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,179 @@
+   * <P>The constructor is private so as to adhere to the rules of the Singleton pattern.</P>
+   */
+  private BriefcaseAnalytics() {
+    hasTrackedStartup = false;
```
This line serves no purpose. It’s already false.

jpknox replied:
I was confusing boolean for Boolean here in terms of behaviour. Boolean defaults to null whereas boolean defaults to false.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcasePreferences.java:
```diff
@@ -181,4 +185,40 @@ public static HttpHost getBriefCaseProxyConnection() {
+   * @return the String representation of the user's UUID.
+   */
+  public static String getUniqueUserID() {
+    String defaultUuidValue = "UUID missing, defaulting to this value";
```
The important noun in this noun phrase is message, not value. It’s a message about a missing UUID.

jpknox replied:
Good point, I'll amend it.

dcbriccetti replied:
I don’t see this change yet. Perhaps you’re still making changes.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcaseAnalytics.java:
```diff
@@ -0,0 +1,179 @@
+  /**
+   * Constructor
+   *
+   * <P>This constructor will instantiate 'brsanthu's {@link GoogleAnalytics} class.</P>
```
The comment is no longer correct.

jpknox replied:
Technically it is, when the class is instantiated it will also instantiate the `GoogleAnalytics` field.

dcbriccetti replied:
Not so. Since googleAnalytics is a class variable (static), it is initialized before any instances of the class are created.

jpknox replied:
I disagree, because googleAnalytics is a static field, it is initialised when the first instance of BriefcaseAnalytics is initialised, BriefcaseAnalytics is a Singleton so the comment on the constructor is technically correct, albeit it's not necessary and it adds noise. 

Your comment was thought provoking and it led me to learn more about Java, keep them coming 👍 .

dcbriccetti replied:
You are mistaken, my new friend, as I demonstrate in the debugger here.

<img width="871" alt="screen shot 2017-08-15 at 12 00 45 pm" src="https://user-images.githubusercontent.com/55099/29331585-cacd1f18-81b1-11e7-9af4-094684f30fa5.png">

It is created when this line is executed from ```MainBriefcaseWindow```:
```  private static BriefcaseAnalytics briefcaseAnalytics = BriefcaseAnalytics.getInstance();```

But it’s moot now because you have removed the comment.

jpknox commented:
PR Updated.

dcbriccetti commented on src/org/opendatakit/briefcase/model/BriefcasePreferences.java:
```diff
@@ -181,4 +185,40 @@ public static HttpHost getBriefCaseProxyConnection() {
+   * Get the user's persisted decision regarding their consent to being tracked.
+   * @return the boolean representation of the user's consent to being tracked.
+   */
+  public static Boolean getBriefcaseTrackingConsentProperty() {
```
This should return ```boolean``` (primitive, not reference). No need to complicate things with the return type suggesting the value could be null.

jpknox replied:
I was against doing this until I found both JCheckedBox.isSelected() and GoogleAnalyticsConfig.setGatherStats() use boolean instead of Boolean. My choice to go with Boolean was because it made persisting it easier because of how it _has_ to be stored as a String, ergo Boolean.toString(). I've read that it's bad to jump from similar type to similar type (Boolean to boolean and vice versa), which I couldn't avoid even though I thought I was doing so. Upon learning about the two previously mentioned signatures I decided it's probably best to keep the complexity of storing it inside the class responsible for storing it, and to keep using boolean everywhere else.

dcbriccetti replied:
Glad we are coming together. A note on autoboxing: performance might be an issue if it happens frequently. Otherwise it’s secondary to type correctness and clarity.

https://docs.oracle.com/javase/tutorial/java/data/autoboxing.html

jpknox commented:
Updated the PR. I've converted the getter/setter of the tracking-consent boolean value to use a primitive boolean instead of a Boolean, to contain the complexity of how BriefcasePreferences needs to store a String, it will continue to convert it to a Boolean before ultimately converting it to a String; the reverse is true when getting it.

I removed a line from the comment for BriefcaseAnalytics' constructor.

dcbriccetti commented:
OK, @lognaturel, I’m happy with this and believe it can be merged. I want to come back and fix the JCheckBoxes (problem that preexisted this change), and once Briefcase is set up for testing, add some tests.

lognaturel commented:
Hurray! Thanks for working closely on this @jpknox and @dcbriccetti! And a huge thanks to @jpknox for taking this on, it's a feature that has been on the wishlist for a very long time.

@yanokwa do you have a tracking ID for this? If you share it in the Slack with @jpknox he can change it when he has a moment before we merge.

yanokwa commented:
I'll update the tracking ID in master as soon as I merge.
